### ISSUE TITLE
Rewrite / rethink I.2

### ISSUE THREAD

**Issue by ghbplayer** [isocpp/CppCoreGuidelines#1695]
> Avoid non-const global variables

I've tried following this advice many times over the years, and it can be very bad advice. We all know the problems with globals, namely that anyone can access them. However, some things are naturally global. Depending on the program, that might include things like network status, logger, various caches, message queues, performance statistics, and so forth. 

When you try to "solve" globals that really need to be global by hiding them and passing references in function calls, you end up with a lot of extra code. When a bit of code needs access, you have to put on your plumber's hat and figure out how to get the data from where you're hiding it to where it's needed. Also, there's no guarantee that some code won't take a reference to this pseudo-global.

In short, you write more code and make it harder to figure out who's using the global. You can grep all files to find out who's using a global, you can't do the same when passing references around everywhere.

There is no benefit to pretending a global is not a global. This applies to Singletons and every other form of hiding globals.

A skilled developer will identify what data/classes/functions are naturally global and which are not. Minimize globals as much as practical and no further.

**Comment by shaneasd** [comment]
It's a question of trade-offs right? If you make something that doesn't need to be global because it's convenient then you encounter the problems documented in I.2. If you instead pass a reference to this object throughout your code then that is a lot more code that doesn't really solve that problem. However if you can make it non-global without that web of references or you can make it const then that is preferable. The rule explicitly acknowledges that there are circumstances where non-const globals are acceptable. It's just that in those circumstances you have to work harder to ensure the object is used correctly (e.g. avoiding data races). I think with all "avoid" rules it's a question of highlighting that there is a cost to violating the rule. It doesn't mean you should never do it. If I code review a piece of code which involves a global which, in my judgement, should not be a global then I want to be able to link the author to this rule as explanation as to why there is a cost to using globals.

It sounds to me like you have the experience to back a strong intuition as to when the trade-off is in favor of using a global and when it isn't but the core guidelines should be for c++ developers of all skill levels. I only disagree with one of your assertions and that is "We all know the problems with globals". I am happy to accept that you know the problems with globals but I assert that many programmers do not. I think for those programmers avoiding non-const globals is the right default and perhaps they should defer to a "skill developer" if considering making something a non-const global.

**Comment by hsutter** [comment]
Editors call: 

>  it can be very bad advice.

We agree there can be reasons to have them, which is why the item is worded "Avoid" (prefer not to use) rather than "Never use."
